### PR TITLE
Extract completion provider crate

### PR DESCRIPTION
**Pull request by rtfeldman**
We will soon need `semantic_index` to be able to use `CompletionProvider`. This is currently impossible due to a cyclic crate dependency, because `CompletionProvider` lives in the `assistant` crate, which depends on `semantic_index`.

This PR breaks the dependency cycle by extracting two crates out of `assistant`: `language_model` and `completion`.

Only one piece of logic changed: [this code](https://github.com/zed-industries/zed/commit/922fcaf5a6076e56890373035b1065b13512546d#diff-3857b3707687a4d585f1200eec4c34a7a079eae8d303b4ce5b4fce46234ace9fR61-R69).
* As of https://github.com/zed-industries/zed/pull/13276, whenever we ask a given completion provider for its available models, OpenAI providers would go and ask the global assistant settings whether the user had configured an `available_models` setting, and if so, return that.
* This PR changes it so that instead of eagerly asking the assistant settings for this info (the new crate must not depend on `assistant`, or else the dependency cycle would be back), OpenAI completion providers now store the user-configured settings as part of their struct, and whenever the settings change, we update the provider.

In theory, this change should not change user-visible behavior...but since it's the only change in this large PR that's more than just moving code around, I'm mentioning it here in case there's an unexpected regression in practice! (cc @amtoaer in case you'd like to try out this branch and verify that the feature is still working the way you expect.)

Release Notes:

- N/A

### CODE REVIEW

File: crates/semantic_index/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ anyhow.workspace = true
 client.workspace = true
 clock.workspace = true
 collections.workspace = true
+completion.workspace = true
```
Comment (rtfeldman) [inline crates/semantic_index/Cargo.toml:25]: I'm adding the dep now in order to make sure a future PR doesn't introduce a cycle between now and then!

Comment (maxdeviant) [reply crates/semantic_index/Cargo.toml:25]: You might need to add a phony reference to it in the code as well, so that `cargo-machete` doesn't complain.

Comment (rtfeldman) [reply crates/semantic_index/Cargo.toml:25]: Yep, didn't realize we ran that on CI...but I'm glad we do! 😄

**Comment by maxdeviant** [comment]
@rtfeldman I fixed up the issues that CI was failing on with the `FakeCompletionProvider`.

The trick is that we need to add `completion as a dev dependency with the `test-support` feature so that we can reference it in tests in another crate:

https://github.com/zed-industries/zed/pull/14823/commits/dcfbfa88e3d7adabbd794faffcb4669e0dda0a4b

**Pull request merged**